### PR TITLE
ref(ts): convert pluginComponentBase and settings to TS

### PR DESCRIPTION
**Pull request by scefali**

### CODE REVIEW

File: src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx [hunk 1]
```diff
@@ -40,44 +51,46 @@ class PluginComponentBase extends React.Component {
...
   }
 
-  onLoadSuccess(callback, ...args) {
-    this.setState(
-      {
-        state: FormState.READY,
-      },
-      callbackWithArgs(callback, ...args)
-    );
+  onLoadSuccess() {
```
Comment (scefali) [inline src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:82]: We don't actually pass a callback in and it was causing TS errors so I just removed that functionality

File: src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx [hunk 2]
```diff
@@ -11,17 +11,28 @@ import {
 } from 'app/actionCreators/indicator';
 import {t} from 'app/locale';
 
-const callbackWithArgs = function (callback, ...args) {
+const callbackWithArgs = function (context: any, callback, ...args) {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:14]:
```suggestion
const callbackWithArgs = function (context: any, callback, ...args: any[]) {
```

File: src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx [hunk 3]
```diff
@@ -11,17 +11,28 @@ import {
...
   if (isFunction(callback)) {
-    callback = callback.bind(this, ...args);
+    callback = callback.bind(context, ...args);
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:16]: as you are touching this file maybe you could use ternary and remove this else? :)

File: src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx [hunk 4]
```diff
@@ -11,17 +11,28 @@ import {
...
   return callback;
 };
 
-class PluginComponentBase extends React.Component {
-  constructor(props, context) {
+type GenericFieldProps = Parameters<typeof GenericField>[0];
+
+type Props = {
+  config?: GenericFieldProps['config'];
+} & GenericFieldProps;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:27]: I believe you forgot to omit the config here 

```suggestion
} & Omit<GenericFieldProps, 'config'>;
```

File: src/sentry/static/sentry/app/plugins/components/settings.tsx [hunk 5]
```diff
@@ -9,9 +9,31 @@ import {t, tct} from 'app/locale';
...
+
+type Props = {
+  organization: Organization;
+  project: Project;
+  plugin: Plugin;
+} & PluginComponentBase['props'];
+
+type State = {
+  fieldList: any[] | null;
+  initialData: any | null;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/plugins/components/settings.tsx:22]: Please, isn't initialData an object?

File: src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx [hunk 6]
```diff
@@ -126,15 +132,18 @@ class PluginComponentBase extends React.Component {
...
+  renderField(props: Omit<GenericFieldProps, 'formState'>) {
     props = {...props};
-    props.formState = this.state.state;
-    props.config = props.config || {};
-    return <GenericField key={props.config.name} {...props} />;
+    const newProps = {
+      ...props,
+      formState: this.state.state,
+    };
+    console.log('config', newProps.config, newProps);
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:145]: I think you might want to remove this console 🙂
```suggestion
```

File: src/sentry/static/sentry/app/plugins/components/settings.tsx [hunk 7]
```diff
@@ -8,10 +8,36 @@ import {parseRepo} from 'app/utils';
...
+  plugin: Plugin;
+} & PluginComponentBase['props'];
+
+type Field = Parameters<typeof PluginComponentBase.prototype.renderField>[0]['config'];
+
+type BackendField = Field & {value?: any; defaultValue?: any};
+
+type State = {
+  fieldList: Field[] | null;
+  initialData: Record<string, any> | null;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/plugins/components/settings.tsx:26]: 👏

File: src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx [hunk 8]
```diff
@@ -11,17 +11,21 @@ import {
...
 
-const callbackWithArgs = function (callback, ...args) {
-  if (isFunction(callback)) {
-    callback = callback.bind(this, ...args);
-  } else {
-    callback = null;
-  }
-  return callback;
+const callbackWithArgs = function (context: any, callback: any, ...args: any) {
+  return isFunction(callback) ? callback.bind(context, ...args) : null;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:15]: 👏

File: src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx [hunk 9]
```diff
@@ -11,17 +11,21 @@ import {
...
 };
 
-class PluginComponentBase extends React.Component {
-  constructor(props, context) {
+type GenericFieldProps = Parameters<typeof GenericField>[0];
+
+type Props = {};
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:20]: don't you need the `config` prop here anymore?

Comment (scefali) [reply src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:20]: @priscilawebdev Nah, it's always going to be passed to `renderField` and doesn't need a default value.

File: src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx (hunk 8 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/bases/pluginComponentBase.tsx:15]:
```suggestion
  return isFunction(callback) ? callback.bind(context, ...args) : undefined;
```

**Comment by priscilawebdev** [comment]
Almost there 😉

**Review by priscilawebdev — APPROVED** [review]
Good Job 🎉

**Pull request merged**